src/System.IO/tests/BinaryWriter/BinaryWriter_WriteByteCharTests.cs

```diff
@@ -0,0 +1,571 @@
+        /// 
+        /// We dont throw on the second read but then throws continuously - note the loop count difference in the 2 loops
+        /// 
+        /// BinaryReader was reverting to its original location instead of advancing. This was changed to skip past the char in the surrogate range.
+        /// The affected method is InternalReadOneChar (IROC). Note that the work here is slightly complicated by the way surrogates are handled by 
+        /// the decoding classes. When IROC calls decoder.GetChars(), if the bytes passed in are surrogates, UnicodeEncoding doesn't report it. 
+        /// charsRead would end up being one value, and since BinaryReader doesn't have the logic telling it exactly how many bytes total to expect, 
+        /// it calls GetChars in a second loop. In that loop, UnicodeEncoding matches up a surrogate pair. If it realizes it's too big for the encoding, 
+        /// then it throws an ArgumentException (chars overflow). This meant that BinaryReader.IROC is advancing past two chars in the surrogate 
+        /// range, which is why the position actually needs to be moved back (but not past the first surrogate char).  
+        /// 
+        /// Note that UnicodeEncoding doesn't always throw when it encounters two successive chars in the surrogate range. The exception 
+        /// encountered here happens if it finds a valid pair but then determines it's too long. If the pair isn't valid (a low then a high), 
+        /// then it returns 0xfffd, which is why BinaryReader.ReadChar needs to do an explicit check. (It always throws when it encounters a surrogate)
+        /// </summary>
+        [Fact]
+        public static void BinaryWriter_WriteCharArrayTest2()
+        {
+
+            MemoryStream mem = new MemoryStream();
+            BinaryWriter writer = new BinaryWriter(mem, Encoding.Unicode);
+
+            // between 55296 <= x < 56319
+
+            // between 56320 <= x < 57343
+            Char[] randomChars = new Char[] { 
+                (Char)55296, (Char)57297, (Char)55513, (Char)56624, (Char)55334, (Char)56957, (Char)55857, 
+                (Char)56355, (Char)56095, (Char)56887, (Char) 56126, (Char) 56735, (Char)55748, (Char)56405,
+                (Char)55787, (Char)56707, (Char) 56300, (Char)56417, (Char)55465, (Char)56944
+            };
+
+            writer.Write(randomChars);
+            mem.Position = 0;
+            BinaryReader reader = new BinaryReader(mem, Encoding.Unicode);
+
+            for (int i = 0; i < 50; i++)
+            {
+                try
+                {
+                    reader.ReadChar();
+
+                    Assert.Equal(1, i);
+                }
+                catch (ArgumentException)
+                {
+                    // I guess sometimes it throws and othertimes, it doesn't?
+                    // Based on the comments above.
```

Comment:
hahaha. We should (at the very least) change this comment to be more sure of itself imo.

Something like ```ArgumentException is sometimes thrown on ReadChar() due to the behavior outlined in the method summary.``` would be far more reassuring.